Dynamic storage opcodes: base instructions (#640)
FuelLabs/fuel-specs: src/fuel-vm/instruction-set.md

  - [`RETD`: Return from context with data](#retd-return-from-context-with-data)
  - [`RVRT`: Revert](#rvrt-revert)
  - [`SMO`: Send message to output](#smo-send-message-out)
  - [`TIME`: Timestamp at height](#time-timestamp-at-height)
  - [`TR`: Transfer coins to contract](#tr-transfer-coins-to-contract)
  - [`TRO`: Transfer coins to output](#tro-transfer-coins-to-output)
  - [Contract State Instructions](#contract-state-instructions)

>     - [`SCWQ`: DEPRECATED State clear sequential slots](#scwq-deprecated-state-clear-sequential-slots)

Reviewer: To better distinguish it from the description of `SCLR`.
```suggestion
    - [`SCWQ`: DEPRECATED State clear sequential 32 bytes slots](#scwq-deprecated-state-clear-sequential-slots)
```

    - [`SRW`: State read word](#srw-state-read-word)
    - [`SRWQ`: DEPRECATED State read sequential 32 byte slots](#srwq-deprecated-state-read-sequential-32-byte-slots)
    - [`SWW`: State write word](#sww-state-write-word)
    - [`SWWQ`: DEPRECATED State write sequential 32 byte slots](#swwq-deprecated-state-write-sequential-32-byte-slots)
    - [`SCLR`: State clear sequential slots](#sclr-state-clear-sequential-slots)

>     - [`SRDD`: Read storage slot](#srdd-read-storage-slot)

Reviewer: To follow the existing convention of using `state` instead of `storage` and following the "State <verb> slot" naming patter. We use "state" also in the caption so it makes sense to stick to this term.
```suggestion
    - [`SRDD`: State read slot](#srdd-state-read-slot)
```

>     - [`SRDI`: Read storage slot immediate](#srdi-read-storage-slot-immediate)

Reviewer: ```suggestion
    - [`SRDI`: State read slot immediate](#srdi-state-read-slot-immediate)
```

>     - [`SWRD`: Write storage slot](#swrd-write-storage-slot)

Reviewer: ```suggestion
    - [`SWRD`: State write slot](#swrd-state-write-slot)
```

>     - [`SWRI`: Write storage slot immediate](#swri-write-storage-slot-immediate)

Reviewer: ```suggestion
    - [`SWRI`: State write slot immediate](#swri-state-write-slot-immediate)
```

>     - [`SUPD`: Update storage slot (partial write)](#supd-update-storage-slot-partial-write)

Reviewer: ```suggestion
    - [`SUPD`: State update slot (partial write)](#supd-state-update-slot-partial-write)
```

>     - [`SUPI`: Update storage slot (partial write) immediate](#supi-update-storage-slot-partial-write-immediate)

Reviewer: ```suggestion
    - [`SUPI`: State update slot (partial write) immediate](#supi-state-update-slot-partial-write-immediate)
```

>     - [`SPLD`: Storage preload](#spld-storage-preload)

Reviewer: ```suggestion
    - [`SPLD`: State preload slot](#spld-state-preload-slot)
```

>     - [`SPCP`: Copy from preloaded storage slot](#spcp-copy-from-preloaded-storage-slot)

Reviewer: Not existing anymore.
```suggestion
```

Panic if:

- `$rA + 32` overflows or `> VM_MAX_RAM`
- `$rB` is a [reserved register](./index.md#semantics)
- `$fp == 0` (in the script context)

Register `$rB` will be set to `false` if any storage slot in the requested range was already unset (default) and `true` if all the slots were set.

### `SRW`: State read word

|             |                                                   |
|-------------|---------------------------------------------------|
| Description | A word is read from the current contract's state. |
| Operation   | ```$rA = STATE[MEM[$rC, 32]][imm * 8, 8];```      |
| Syntax      | `srw $rA, $rB, $rC, imm`                          |
| Encoding    | `0x00 rA rB rC imm`                               |
| Effects     | Storage read                                      |
| Notes       | Returns zero if the state element does not exist. |

Panic if:

- `$rA` is a [reserved register](./index.md#semantics)
- `$rB` is a [reserved register](./index.md#semantics)
- `$rC + 32` overflows or `> VM_MAX_RAM`
- `$fp == 0` (in the script context)
- `len(STATE[MEM[$rC, 32]]) < imm * 8 + 8` (the storage slot doesn't have enough data)

> Register `$rB` will be set to `false` if the requested slot is unset (default) and `true` if it's set.

Author: Clarified in 9dc546a

|-------------|--------------------------------------------------------------------------------------------------------|
| Description | Read storage slot contents to memory. Allows partial reads as well.                                    |
| Operation   | `MEM[$rA, $rD] = STATE[MEM[$rB, 32]][$rC, $rD]`                                                        |
| Syntax      | `srdd $rA, $rB, $rC, $rD`                                                                              |
| Encoding    | `0x00 rA rB rC rD`                                                                                     |
| Effects     | Storage read                                                                                           |
| Notes       | Gas is always charged for a full slot read.                                                            |
Panic if:

- `$rB + 32` overflows or `> VM_MAX_RAM`
- `$rA + $rD` overflows or `> VM_MAX_RAM`
- The memory range `MEM[$rA, $rD]` does not pass [ownership check](./index.md#ownership)
- `$rC + $rD` overflows or `> len(STATE[MEM[$rB, 32]])`
- `$fp == 0` (in the script context)

> Register `$err` will be set to `1` if the slot did not exist, and `0` otherwise. If the slot didn't exist, memory is not modified.

Author: Clarified in 9dc546a

### `SRDI`: State read slot immediate

|             |                                                                                                        |
|-------------|--------------------------------------------------------------------------------------------------------|
| Description | Read storage slot contents to memory. Allows partial reads as well.                                    |
| Operation   | `MEM[$rA, imm] = STATE[MEM[$rB, 32]][$rC, imm]`                                                        |
| Syntax      | `srdi $rA, $rB, $rC, imm`                                                                              |
| Encoding    | `0x00 rA rB rC imm`                                                                                    |
| Effects     | Storage read                                                                                           |
| Notes       | Gas is always charged for a full slot read.                                                            |

Panic if:

- `$rB + 32` overflows or `> VM_MAX_RAM`
- `$rA + imm` overflows or `> VM_MAX_RAM`
- The memory range `MEM[$rA, imm]` does not pass [ownership check](./index.md#ownership)

> - `$rC + imm` overflows or `> len(STATE[MEM[$rC, 32]])`

Author: Fixed in 36b4bf3

This modifies the `balanceRoot` field of the appropriate output(s).

## Contract State Instructions

> ### `SCWQ`: DEPRECATED State clear sequential slots

Reviewer: ```suggestion
### `SCWQ`: DEPRECATED State clear sequential 32 bytes slots
```

Panic if:

- `$rA + 32` overflows or `> VM_MAX_RAM`
- `$rC + 32 * $rD` overflows or `> VM_MAX_RAM`
- `$rB` is a [reserved register](./index.md#semantics)
- `$fp == 0` (in the script context)

Register `$rB` will be set to the number of storage slots that were previously unset, and were set by this operation.

### `SCLR`: State clear sequential slots

|             |                                                                                    |
|-------------|------------------------------------------------------------------------------------|
| Description | A sequential series of slots cleared from the current contract's state.            |
| Operation   | ```STATE[MEM[$rA, 32], $rB] = None;```                                             |
| Syntax      | `sclr $rA, $rB`                                                                    |
| Encoding    | `0x00 rA rB - -`                                                                   |
| Notes       |                                                                                    |

Panic if:

- `$rA + 32` overflows or `> VM_MAX_RAM`
- `$fp == 0` (in the script context)

> ### `SRDD`: Read storage slot

Reviewer: ```suggestion
### `SRDD`: State read slot
```

|-------------|--------------------------------------------------------------------------------------------------------|
| Description | Read storage slot contents to memory. Allows partial reads as well.                                    |
| Operation   | `MEM[$rA, $rD] = STATE[MEM[$rB, 32]][$rC, $rD]`                                                        |
| Syntax      | `srdd $rA, $rB, $rC, $rD`                                                                              |
| Encoding    | `0x00 rA rB rC rD`                                                                                     |
| Effects     | Storage read                                                                                           |
| Notes       | Gas is always charged for a full slot read.                                                            |
Panic if:

- `$rB + 32` overflows or `> VM_MAX_RAM`
- `$rA + $rD` overflows or `> VM_MAX_RAM`
- The memory range `MEM[$rA, $rD]` does not pass [ownership check](./index.md#ownership)
- `$rC + $rD` overflows or `> len(STATE[MEM[$rB, 32]])`
- `$fp == 0` (in the script context)

> Register `$err` will be set to `1` is the slot did not exist, and `0` otherwise. If the slot didn't exist, memory is not modified.

Reviewer: ```suggestion
Register `$err` will be set to `1` if the slot did not exist, and `0` otherwise. If the slot didn't exist, memory is not modified.
```

> Register `$err` will be set to `1` is the slot did not exist, and `0` otherwise. If the slot didn't exist, memory is not modified.

Reviewer: ```suggestion
Register `$err` will be set to `1` if the slot did not exist, and `0` otherwise. If the slot didn't exist, memory is not modified.
```

> ### `SRDI`: Read storage slot immediate

Reviewer: ```suggestion
### `SRDI`: State read slot immediate
```

|-------------|--------------------------------------------------------------------------------------------------------|
| Description | Read storage slot contents to memory. Allows partial reads as well.                                    |
| Operation   | `MEM[$rA, imm] = STATE[MEM[$rB, 32]][$rC, imm]`                                                        |
| Syntax      | `srdi $rA, $rB, $rC, imm`                                                                              |
| Encoding    | `0x00 rA rB rC imm`                                                                                    |
| Effects     | Storage read                                                                                           |
| Notes       | Gas is always charged for a full slot read.                                                            |

Panic if:

- `$rB + 32` overflows or `> VM_MAX_RAM`
- `$rA + imm` overflows or `> VM_MAX_RAM`
- The memory range `MEM[$rA, imm]` does not pass [ownership check](./index.md#ownership)
- `$rC + imm` overflows or `> len(STATE[MEM[$rC, 32]])`
- `$fp == 0` (in the script context)

> Register `$err` will be set to `1` is the slot did not exist, and `0` otherwise. If the slot didn't exist, memory is not modified.

Reviewer: ```suggestion
Register `$err` will be set to `1` if the slot did not exist, and `0` otherwise. If the slot didn't exist, memory is not modified.
```

> ### `SWRD`: Write storage slot

Reviewer: ```suggestion
### `SWRD`: State write slot
```

|             |                                                                                                        |
|-------------|--------------------------------------------------------------------------------------------------------|
| Description | Write storage slot from a memory buffer.                                                               |
| Operation   | `STATE[MEM[$rA, 32]] = MEM[$rB, $rC]`                                                                  |
| Syntax      | `swrd $rA, $rB, $rC`                                                                                   |
| Encoding    | `0x00 rA rB rC -`                                                                                      |
| Effects     | Storage write                                                                                          |
| Notes       |                                                                                                        |

Panic if:

- `$rA + 32` overflows or `> VM_MAX_RAM`
- `$rB + $rC` overflows or `> VM_MAX_RAM`
- `$rC > MAX_STORAGE_SLOT_SIZE`
- `$fp == 0` (in the script context)

> ### `SWRI`: Write storage slot immediate

Reviewer: ```suggestion
### `SWRI`: State write slot immediate
```

| Operation   | `key=MEM[$rA, 32]; tmp = STATE[key]; tmp[$rC==MAX?len(tmp):$rC, $rD] = MEM[$rB, $rD]; STATE[key] = tmp`|
| Syntax      | `supd $rA, $rB, $rC, $rD`                                                                              |
| Encoding    | `0x00 rA rB rC rD`                                                                                     |
| Effects     | Storage read and write                                                                                 |
| Notes       | Charges gas for full read and write. Writing past the end extends the slot, but offset must be valid.  |

An unoccupied slot is treated as zero-length for the purposes of the read. Passing in `u64::MAX` in `$rC` will cause the write to happen at the end of the slot, without needing to read the slot length first.

Panic if:

- `$rA + 32` overflows or `> VM_MAX_RAM`
- `$rB + $rD` overflows or `> VM_MAX_RAM`
- `$rC + $rD` overflows or `> MAX_STORAGE_SLOT_SIZE` (if `$rC == u64::MAX`, read `$rC` as `len(STATE[MEM[$rA, 32]])` instead)
- `$rC` overflows or `> len(STATE[MEM[$rA, 32]])` (except when `$rC == u64::MAX`)
- `$fp == 0` (in the script context)

> ### `SUPI`: Update storage slot (partial write) immediate

Reviewer: ```suggestion
### `SUPI`: State update slot (partial write) immediate
```

| Operation   | `key=MEM[$rA, 32]; tmp = STATE[key]; tmp[$rC==MAX?len(tmp):$rC, imm] = MEM[$rB, imm]; STATE[key] = tmp`|
| Syntax      | `supi $rA, $rB, $rC, imm`                                                                              |
| Encoding    | `0x00 rA rB rC imm`                                                                                    |
| Effects     | Storage read and write                                                                                 |
| Notes       | Charges gas for full read and write. Writing past the end extends the slot, but offset must be valid.  |

An unoccupied slot is treated as zero-length for the purposes of the read. Passing in `u64::MAX` in `$rC` will cause the write to happen at the end of the slot, without needing to read the slot length first.

Panic if:

- `$rA + 32` overflows or `> VM_MAX_RAM`
- `$rB + imm` overflows or `> VM_MAX_RAM`
- `$rC + imm` overflows or `> MAX_STORAGE_SLOT_SIZE` (if `$rC == u64::MAX`, read `$rC` as `len(STATE[MEM[$rA, 32]])` instead)
- `$rC` overflows or `> len(STATE[MEM[$rA, 32]])` (except when `$rC == u64::MAX`)
- `$fp == 0` (in the script context)

> ### `SPLD`: Storage preload

Reviewer: ```suggestion
### `SPLD`: State preload slot
```

|             |                                                                                                        |
|-------------|--------------------------------------------------------------------------------------------------------|

> | Description | Preload a storage slot to a staging area, returning its length.                                        |

Reviewer: ```suggestion
| Description | Preload a storage slot to hot storage, returning its length.                                        |
```

> | Operation   | `STAGING = STATE[MEM[$rB, 32]]; $rA = len(STAGING)`                                                    |

Reviewer 2: Is the staging area unique to each call frame? Seems like there could be some confusing behaviors if there is cross contract contention over the staging area:

ie. 

Contract A: PRELOAD
Contract A: CALL Contract B
Contract B: PRELOAD (with 50% probability)
Contract B: RETURN
Contract A: COPY PRELOAD

Well structured programs shouldn't preload and then leave room for others to change the staging area before using the value, so this is likely a very niche issue. However, it could lead to very unpredictable behavior. Might be worth documenting this issue if the overhead of isolating the staging area to each callframe is too high.

Reviewer 3: The implementation zeroes staging area before the `CALL` and after `RET`, `RETD` opcodes

| Syntax      | `spld $rA, $rB`                                                                                        |
| Encoding    | `0x00 rA rB - -`                                                                                       |
| Effects     | Storage read                                                                                           |
| Notes       | Charges gas for a full read.                                                                           |

Panic if:

- `$rB + 32` overflows or `> VM_MAX_RAM`
- `$rA` is a [reserved register](./index.md#semantics).
- `$fp == 0` (in the script context)

> If the slot doesn't exist, sets `$rA = 0` and `$err = 1`, leaving the staging area unchanged. Otherwise `$rA` is set to the length of the slot, and `$err` to `0`.

Reviewer: ```suggestion
If the slot doesn't exist, sets `$rA = 0` and `$err = 1`. Otherwise `$rA` is set to the length of the slot, and `$err` to `0`.
```

|-------------|--------------------------------------------------------------------------------------------------------|
| Description | Copy memory from preloaded slot to main memory.                                                        |
| Operation   | `MEM[$rA, $rC+imm] = STAGING[$rB, $rC+imm]                                                             |
| Syntax      | `spcp $rA, $rB, $rC`                                                                                   |
| Encoding    | `0x00 rA rB rC imm`                                                                                    |
| Effects     | Storage read                                                                                           |
| Notes       |                                                                                                        |

Panic if:

- `$rA + $rC + imm` overflows or `> VM_MAX_RAM`
- The memory range `MEM[$rA, $rC + imm]` does not pass [ownership check](./index.md#ownership)
- `$rB + $rC + imm` overflows or `> len(STAGING)`
- `$fp == 0` (in the script context)

> 

Reviewer: ```suggestion
```
